shared/actions/signup.js

```diff
@@ -1,366 +1,304 @@
-  return (dispatch, getState) => {
-    // TODO careful here since this will not be sync on a remote component!
-    const phase: string = getState().signup.phase
-    dispatch(navigateAppend([phase], [loginTab, 'signup']))
+const checkInviteCode = (action: SignupGen.CheckInviteCodePayload) =>
+  Saga.call(
+    RPCTypes.signupCheckInvitationCodeRpcPromise,
+    {invitationCode: action.payload.inviteCode},
+    Constants.waitingKey
+  )
+const checkInviteCodeSuccess = (_, action: SignupGen.CheckInviteCodePayload) =>
+  Saga.sequentially([
+    Saga.put(SignupGen.createCheckInviteCodeDone({inviteCode: action.payload.inviteCode})),
+    Saga.put(navigateTo([loginTab, 'signup', 'usernameAndEmail'])),
+  ])
+const checkInviteCodeError = () =>
+  Saga.put(SignupGen.createCheckInviteCodeDoneError({errorText: "Sorry, that's not a valid invite code."}))
+
+const requestAutoInvite = () =>
+  Saga.call(RPCTypes.signupGetInvitationCodeRpcPromise, undefined, Constants.waitingKey)
+const requestAutoInviteSuccess = (inviteCode: string) =>
+  Saga.put(SignupGen.createCheckInviteCode({inviteCode}))
+const requestAutoInviteError = () => Saga.put(navigateTo([loginTab, 'signup', 'inviteCode']))
+
+const requestInvite = (action: SignupGen.RequestInvitePayload) => {
+  const {email, name} = action.payload
+  const emailError = isValidEmail(email)
+  const nameError = isValidName(name)
+  if (emailError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError, name, nameError: ''}))
+  }
+  if (nameError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError: '', name, nameError}))
   }
+
+  return Saga.call(
+    RPCTypes.signupInviteRequestRpcPromise,
+    {email: email, fullname: name, notes: 'Requested through GUI app'},
+    Constants.waitingKey
+  )
 }
 
-function startRequestInvite() {
-  return (dispatch: Dispatch) => {
-    dispatch(SignupGen.createStartRequestInvite())
-    dispatch(nextPhase())
+const requestInviteSuccess = (
+  result: SignupGen.RequestInviteDonePayload | void,
+  action: SignupGen.RequestInvitePayload
+) => {
+  // rpc returns undefined, dispatches above on error return the type
+  if (result) {
+    return
   }
+  const {email, name} = action.payload
+  return Saga.sequentially([
+    Saga.put(SignupGen.createRequestInviteDone({email, name})),
+    Saga.put(navigateAppend(['requestInviteSuccess'], [loginTab, 'signup'])),
+  ])
+}
+
+const requestInviteError = (err, action: SignupGen.RequestInvitePayload) => {
+  const {email, name} = action.payload
+  return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError: err, name, nameError: ''}))
 }
 
-function checkInviteCodeThenNextPhase(inviteCode: string) {
-  return (dispatch: Dispatch) => {
-    const p = new Promise((resolve, reject) => {
-      dispatch(SignupGen.createCheckInviteCode({inviteCode}))
+const checkUsernameEmail = (action: SignupGen.CheckUsernameEmailPayload) => {
+  const {email, username} = action.payload
```

Comment:
better typing: fix `import type {RPCError}` to `import {RPCError}` (since it's a class)

then

```
  const {email, username} = action.payload

  if (err instanceof RPCError) {
    return Saga.put(
        ... usernameError: `Sorry, there was a problem: ${err.desc}`, ...
    )
  }

  if (err.email) {
    return Saga.put(
        ... emailError: err.email, ...
    )
  } else if (err.username) {
    return Saga.put(
        ... usernameError: err.username, ...
    )
  } else {
     // put some default error or something?
  }
```

should type-check